## Table of Content

1. **[Built-in Functions](#builtin)**

2. **[User Defined Functions](#user)**
    - 2.1 - **[User defined functions without arguments](#without_arg)**
    - 2.2 - **[User defined functions with arguments](#with_arg)**

3. **[Lambda Functions](#lambda)**

4. **[Overview of map, reduce and filter functions](#overview)**
    - 5.1 - [Function map()](#map)
    - 5.2 - [Function filter()](#filter)
    - 5.3 - [Function reduce()](#reduce)





<a id="builtin"> </a>
## 1. Built-in Functions

Some exapmles are: 

<table border="2">
<tr>
<th> Function </th>
<th> Description </th>
</tr>
<tr>
<th> abs()</th>
<th> It gives the absolute of a numeric value </th>
<tr>
<tr>
<th> sum()</th>
<th> It gives the sum of a numeric values values </th>
<tr>
<tr>
<th> len()</th>
<th> It gives the sum of a numeric values values </th>
<tr>   
<tr>
<th> apply()</th>
<th> applied both to series and dataframes where function can be applied both series and individual elements based on the type of function provide </th>
<tr> 
<tr>
<th> map()</th>
<th> only works on a pandas series where type of operation to be applied depends on argument passed as a function, dictionary or a list </th>
<tr>         
</table>    


<a id="user"> </a>
## 2. User Defined Functions

There are situations where in built functions are not availabe in a library. In such cases, we do the computations by using formulas/concepts

Consider the following user defined function:

Now, if we invoke this function:

Here the 'sample_parameters' is called argument.

<a id="without_arg"> </a>
## 2.1 User defined functions without arguments

Once a function is defined, it can be called anytime, without running the function again. Generally, a function returns a value or prints a value.  

Let us see a few examples of user defined functions.

**1. We shall write a user defined function without argument to find the twice of a number.**

In [1]:
def twice():
    a = int(input('enter a number: '))
    print(a*2)

In [2]:
# In this function, no argument is passed
# call the function
twice()

10


<a id="with_arg"> </a>
## 2.2 User defined functions with arguments

Most of the function will either `return` a value or `print` an output. Take a note of this in the examples considered below.

**1. A user defined function to get the average of all numbers in a sequence.**

In [3]:
def average(x):
    count = 0
    total = 0
    for i in x:
        count += 1
        total += i
    avg = total/count
    return avg

In [4]:
x = (44, 45, -56, 44, 32,-54, 44, 56, 44, 87, 44, 56)

In [5]:
# call the function   
average(x)

32.166666666666664

The average of the number is 32.1667

If the function requires an argument to be passed and we do not provide it, then an error with be shown.

In [6]:
# call the function   
average()

TypeError: average() missing 1 required positional argument: 'x'

### `*args`

- *args in function definitions in python is used to pass n number of arguments to a function i.e. there is no need to define the number of arguments at the time of creating the function. 

- Syntax: Use the symbol * to take in n number of arguments. It is often used with the word 'args', however we can use any other  word. 

- For example : we want to make a multiply function that takes any number of arguments and able to multiply them all together. It can be done using *args.

In [ ]:
# def multiply_func(*args): 
#     product=1
#     for arg in args: 
#         product*=arg
#     return product

In [36]:
# * indicates that number of arguments can be taken
def Multiple(*args):  
    for parameter in args:  
        print (parameter) 

In [37]:
#call the function
Multiple("Lets","learn","Python","for","Data","Science") 

Lets
learn
Python
for
Data
Science


### ` **kwargs`

- A keyword argument is where you provide a name to the variable as you pass it into the function. For example, func('Sex'=Female) 
- The special syntax **kwargs in function definitions in python is used to pass a keyworded, variable-length argument list. We use the name kwargs with the double star. The reason is because the double star allows us to pass through keyword arguments (and any number of them). We can use any other name instead of kwargs. 
- One can think of the kwargs as being a dictionary that maps each keyword to the value that we pass alongside it.

In [ ]:
# def full_name(**kwargs): 
#     for key, value in kwargs.items():
#         print ("%s == %s" %(key, value))
# full_name(first ='Great', last='Learning') 

In [40]:
def function(**kwargs):
    for a,b in kwargs.items():
        print(a,"==",b)

In [41]:
# call the function
a=function(A = "First letter", B = "Second letter", C = "Thirs letter")

A == First letter
B == Second letter
C == Thirs letter


<a id = 'lambda' > </a>
## 3. Lambda Functions

Let us look at some examples using on lambda functions.

**1. Obtain the minimum among two numbers**

In [55]:
min_value = (lambda x, y: x if x < y else y)
min_value(34, 78)

34

**2.  Check whether two strings are equal**

In [56]:
result = (lambda x, y: 'True' if x == y else 'False')
result("Data", "DaTa")

'False'

**3. Demonstrating wrong usuage of variable name**

In [57]:
# multiplication - wrong usage of variable
x = lambda a, c : a + b
print(x(5, 6))

NameError: name 'b' is not defined

**If variable a and c are used, then the operations should be on a and c.**

<a id='overview'> </a>
## 4. Overview of map, reduce and filter functions

<a id='map'></a>
### 5.1 Function map()

**1. A test was conducted in a class, the list of all the scores is given. Tha scores are out of 45. Obtain the percentage of each score using lambda function.**
             
             marks = [ 20, 23, 34, 45, 33, 10, 43, 23, 27, 34, 37, 44, 42, 43 ]


In [58]:
marks = [ 20, 23, 34, 45, 33, 10, 43, 23, 27, 34, 37, 44, 42, 43 ]

print(list(map(lambda x, y=100 :x*y/45 , marks)))

[44.44444444444444, 51.111111111111114, 75.55555555555556, 100.0, 73.33333333333333, 22.22222222222222, 95.55555555555556, 51.111111111111114, 60.0, 75.55555555555556, 82.22222222222223, 97.77777777777777, 93.33333333333333, 95.55555555555556]


The output is the percentage scores.

**2. Create two lists and add corresponding elements of each list**

In [59]:
# iteration using list
list_a = [30, 35, 40]
list_b = [12, 5, 70]
  
print(list(map(lambda x, y: x + y, list_a, list_b)))

[42, 40, 110]


<a id='filter'></a>
### 5.2 Function filter()

**1. To obtain the  sum of cubes of odd number from 1 to 50**

In [3]:
# sum of cubes of odd numers from 1 to 50

# user defined function to obtain the cube
def cube(x):
    return(x*x*x)

# user defined function to obtain odd numbers from a list
def isodd(x):
    return(x%2 == 1)

# create a list of all cubes
list_cubes = list(map(cube,filter(isodd,range(1,50))))

# sum of all elements in the in the list_cube
print(list_cubes)
sum(list_cubes)

[1, 27, 125, 343, 729, 1331, 2197, 3375, 4913, 6859, 9261, 12167, 15625, 19683, 24389, 29791, 35937, 42875, 50653, 59319, 68921, 79507, 91125, 103823, 117649]


780625

**2. To obtain the odd numbers from a list**

In [2]:
numbers = [121, 443, 477, 125, 34, 655, 458, 141, 363, 125, 456 , 578, 123]

# function that filters the numbers
def oddnumber(in_num):
    if(in_num % 2) == 0:
        return True
    else:
        return False

# Use filter() to remove odd numbers
out_filter_number = filter(oddnumber, numbers)

print("Type of filter object: ", type(out_filter_number))
print("Even number is as follows: ", list(out_filter_number))

Type of filter object:  <class 'filter'>
Even number is as follows:  [34, 458, 456, 578]


<a id='reduce'></a>
### 4.3 Function reduce()

**1. List of  sales in a shop are noted. Obtain the total sales**

In [4]:
# import reduce function from functools
from functools import reduce

sale = [14565,65460,74450,54546,23245,14543,84355,54566]
total_sale = reduce(lambda a, b: a + b, sale)
print(total_sale)


385730


**Thus we may say the total sale is 385730 INR.**